libyul/optimiser/RedundantStoreEliminator.cpp

```diff
@@ -0,0 +1,447 @@
+	Expression const oneLiteral{Literal{{}, LiteralKind::Number, YulString{"1"}, {}}};
+	Expression const thirtyTwoLiteral{Literal{{}, LiteralKind::Number, YulString{"32"}, {}}};
+	values[YulString{zero}] = AssignedValue{&zeroLiteral, {}};
+	values[YulString{one}] = AssignedValue{&oneLiteral, {}};
+	values[YulString{thirtyTwo}] = AssignedValue{&thirtyTwoLiteral, {}};
+
+	bool const ignoreMemory = MSizeFinder::containsMSize(_context.dialect, _ast);
+	RedundantStoreEliminator rse{
+		_context.dialect,
+		functionSideEffects,
+		ControlFlowSideEffectsCollector{_context.dialect, _ast}.functionSideEffectsNamed(),
+		values,
+		ignoreMemory
+	};
+	rse(_ast);
+	rse.changeUndecidedTo(State::Unused, Location::Memory);
+	rse.changeUndecidedTo(State::Used, Location::Storage);
+	rse.scheduleUnusedForDeletion();
+
+	StatementRemover remover(rse.m_pendingRemovals);
+	remover(_ast);
+}
+
+void RedundantStoreEliminator::operator()(FunctionCall const& _functionCall)
+{
+	RedundantStoreBase::operator()(_functionCall);
+
+	for (Operation const& op: operationsFromFunctionCall(_functionCall))
+		applyOperation(op);
+
+	ControlFlowSideEffects sideEffects;
+	if (auto builtin = m_dialect.builtin(_functionCall.functionName.name))
+		sideEffects = builtin->controlFlowSideEffects;
+	else
+		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
+
+	if (!sideEffects.canContinue)
+	{
+		changeUndecidedTo(State::Unused, Location::Memory);
+
+		changeUndecidedTo(
+			sideEffects.canTerminate ?
+			State::Used :
+			State::Unused,
+			Location::Storage
+		);
+	}
+}
+
+void RedundantStoreEliminator::operator()(FunctionDefinition const& _functionDefinition)
+{
+	ScopedSaveAndRestore storeOperations(m_storeOperations, {});
+	RedundantStoreBase::operator()(_functionDefinition);
+}
+
+
+void RedundantStoreEliminator::operator()(Leave const&)
+{
+	changeUndecidedTo(State::Used);
+}
+
+void RedundantStoreEliminator::visit(Statement const& _statement)
+{
+	using evmasm::Instruction;
+
+	RedundantStoreBase::visit(_statement);
+
+	auto const* exprStatement = get_if<ExpressionStatement>(&_statement);
+	if (!exprStatement)
+		return;
+
+	FunctionCall const* funCall = get_if<FunctionCall>(&exprStatement->expression);
+	if (!funCall)
+		return;
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, funCall->functionName.name);
+	if (!instruction)
+		return;
+
+	if (!ranges::all_of(funCall->arguments, [](Expression const& _expr) -> bool {
+		return get_if<Identifier>(&_expr) || get_if<Literal>(&_expr);
+	}))
+		return;
+
+	if (
+		*instruction == Instruction::SSTORE ||
+		(!m_ignoreMemory && (
+			*instruction == Instruction::EXTCODECOPY ||
+			*instruction == Instruction::CODECOPY ||
+			*instruction == Instruction::CALLDATACOPY ||
+			*instruction == Instruction::RETURNDATACOPY ||
+			*instruction == Instruction::MSTORE ||
+			*instruction == Instruction::MSTORE8
+		))
+	)
+	{
+		m_stores[YulString{}].insert({&_statement, State::Undecided});
```

Comment:
It took me a while to understand what this `m_stores` is about, especially given that it's spread over two classes and `m_storeOperations` looks kinda similar at a glance. It's probably because the big picture of what the code is doing was not immediately clear to me and I had to first piece it together from smaller details. I think this could use a bit more documentation and some things could be named more explicitly. 

As I understand it now, `RedundantStoreBase` does not add anything to `m_stores` on its own. Instead it lets the derived class add the statements it wants to track. For each statement it lets the derived class tell it whether the value was ever accessed and it manages that state by merging it across `if` branches, loop cycles, etc. as it walks the AST. For me it would have been very helpful if I got this info upfront in the class docstring.

As for names, I think that something like `StoredValueAccessTracker` would be more informative than `RedundantStoreBase`. `State` could also be `ValueState` or maybe `StoredValueAccess` (neither is great but yeah, coming up with a good name for this one is hard). Also, maybe making `m_stores` private and adding some helpers with more meaningful names like `trackStatement()`, `markAsUsed()`, etc. would make the code easier to understand when you see it used in the derived class.